src/librustc_trait_selection/traits/select.rs

```diff
@@ -1952,22 +1964,23 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         let is_global =
             |cand: &ty::PolyTraitRef<'_>| cand.is_global() && !cand.has_late_bound_regions();
 
+        // (*) Prefer `BuiltinCandidate { has_nested: false }` and `DiscriminantKindCandidate`
+        // to anything else.
+        //
+        // This is a fix for #53123 and prevents winnowing from accidentally extending the
+        // lifetime of a variable.
         match other.candidate {
-            // Prefer `BuiltinCandidate { has_nested: false }` to anything else.
-            // This is a fix for #53123 and prevents winnowing from accidentally extending the
-            // lifetime of a variable.
-            BuiltinCandidate { has_nested: false } => true,
+            // (*)
+            BuiltinCandidate { has_nested: false } | DiscriminantKindCandidate => true,
```

Comment:
Not sure this is correct. Maybe `BuiltinCandidate` should be repurposed instead of adding `DiscriminantKindCandidate`?

Reply:
While they might be unified, `DiscriminantKind` is handled differently than the other builtins, so I prefer keeping them split, even if they need the same information.

Reply:
cc @nikomatsakis

Reply:
Hmm, I'm not sure which is the best setup. It seems like we have 3 fairly distinct cases:

* `BuiltinCandidate { has_nested: true }` -- used for copy, clone, and `Sized`, though in slightly different ways between those traits
* `BuiltinCandidate { has_nested: false }` -- used for `Unpin` with *movable generator types* specifically
* `AutoImplCandidate` -- used for other auto traits
* `DiscriminantKind` -- for the discriminant trait we are adding now

Of these, `BuiltinCandidate { has_nested: false }` is the "odd one out", because it applies not only to a specific trait, but to a specific *self type* for that trait. Are we actually exposing "movable generators" anywhere in the language? I don't think so, and I don't know that we have any real plans to do so either, right? I wonder if we should simplify things by removing them.

Ignoring `has_nested: false` case for the moment: it seems like the main piece of code that wants to discriminate between traits is the projection code?

One option that might be nice is to have

```rust
enum CandidateKind {
    ...
    BuiltinCandidate(BuiltinCandidateKind)
}

enum BuiltinCandidateKind {
    AutoTrait,

    SizedTrait,
    CopyTrait,
    CloneTrait,

    DiscriminantKindTrait
}
```

Currently, in select.rs anyway, those built-in candidates each behave "kind of" differently, and are grouped in the  groups I showed above. But I suspect that the *select.rs* handling for all of these cases could be combined, at least if we added a limitation on auto traits that they can never have any predicates (true for all known auto traits, and at least *somewhat* required for soundness around coinduction anyhow). In each case the idea would be that you extract some set of types for the "recursive obligations" (sometimes empty, as in the case of `DiscriminantKind`).

The project.rs handling meanwhile could use these types to pick out the right case.

This leaves the question of how to handle `Unpin` -- we could certainly add a special case tothe list above, like `UnpinTraitMovableGenerator` and it would fit ok I suppose.

This seems like it might be a nice refactoring, but it also feels like the majority of this refactoring is orthogonal from this PR. But it feels like it might be nice to do beforehand. @lcnr if you're interested I could try to describe what I have in mind in more detail.

Reply:
I am very interested!

>  This seems like it might be a nice refactoring, but it also feels like the majority of this refactoring is orthogonal from this PR. But it feels like it might be nice to do beforehand.

:thinking: I don't fully know how much this PR would be improved by this,
but I kind of wish that we merge this first.